### PR TITLE
Fix the problem of rewriten null_flag of nullable column

### LINKED ISSUE #3252: Query result error
### Steps to reproduce the behavior (Required)
1. DROP DATABASE IF EXISTS d2;
CREATE DATABASE d2;
USE d2;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE IF NOT EXISTS t0 (c_0_0 DECIMAL(7, 5)  NOT NULL ,c_0_1 DECIMAL(11, 5)  NULL ,c_0_2 TINYINT NULL ,c_0_3 LARGEINT NOT NULL ,c_0_4 DECIMAL(2, 1)  NOT NULL ,c_0_5 SMALLINT NULL ,c_0_6 SMALLINT NULL ,c_0_7 DECIMAL(25, 3)  NULL ,c_0_8 DATETIME NULL ,c_0_9 VARCHAR(11) NOT NULL ,c_0_10 INT NOT NULL ,c_0_11 VARCHAR(31) NOT NULL ,c_0_12 BOOLEAN REPLACE NULL ,c_0_13 BOOLEAN REPLACE NOT NULL ,c_0_14 FLOAT SUM NULL ,c_0_15 HLL HLL_UNION NULL ,c_0_16 DATE REPLACE NOT NULL ,c_0_17 HLL HLL_UNION NOT NULL ,c_0_18 FLOAT MAX NULL ) AGGREGATE KEY (c_0_0,c_0_1,c_0_2,c_0_3,c_0_4,c_0_5,c_0_6,c_0_7,c_0_8,c_0_9,c_0_10,c_0_11) DISTRIBUTED BY HASH (c_0_10,c_0_1,c_0_8) properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t1 (c_1_0 TINYINT NULL ,c_1_1 DECIMAL(1, 0)  NULL ,c_1_2 DATETIME NULL ,c_1_3 DECIMAL(20, 3)  NOT NULL ,c_1_4 SMALLINT NULL ,c_1_5 DECIMAL(24, 13)  NOT NULL ,c_1_6 INT NOT NULL ,c_1_7 DECIMAL(31, 12)  NULL ,c_1_8 BOOLEAN NOT NULL ,c_1_9 BOOLEAN NULL ,c_1_10 TINYINT NOT NULL ,c_1_11 HLL HLL_UNION NULL ,c_1_12 DATETIME REPLACE NOT NULL ,c_1_13 FLOAT SUM NULL ,c_1_14 BOOLEAN REPLACE NOT NULL ,c_1_15 DOUBLE MIN NOT NULL ,c_1_16 TINYINT SUM NULL ) AGGREGATE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4,c_1_5,c_1_6,c_1_7,c_1_8,c_1_9,c_1_10) DISTRIBUTED BY HASH (c_1_10,c_1_3,c_1_6,c_1_1,c_1_8,c_1_2,c_1_5,c_1_0) properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t2 (c_2_0 DATE NOT NULL ,c_2_1 CHAR(31) NULL ,c_2_2 DECIMAL(35, 18)  NULL ,c_2_3 BOOLEAN NULL ,c_2_4 DATE NOT NULL ) DUPLICATE KEY (c_2_0,c_2_1) DISTRIBUTED BY HASH (c_2_1,c_2_0) properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t3 (c_3_0 DECIMAL(11, 2)  NOT NULL ,c_3_1 DECIMAL(17, 15)  NOT NULL ,c_3_2 BOOLEAN NOT NULL ,c_3_3 BOOLEAN NULL ,c_3_4 SMALLINT NOT NULL ,c_3_5 VARCHAR(1) NOT NULL ,c_3_6 BOOLEAN NOT NULL ,c_3_7 DECIMAL(7, 0)  NULL ,c_3_8 BOOLEAN NULL ,c_3_9 BOOLEAN NULL ,c_3_10 VARCHAR(1) NOT NULL ,c_3_11 VARCHAR(31) NOT NULL ,c_3_12 BOOLEAN NULL ) DUPLICATE KEY (c_3_0,c_3_1,c_3_2,c_3_3,c_3_4,c_3_5,c_3_6,c_3_7,c_3_8,c_3_9,c_3_10,c_3_11) DISTRIBUTED BY HASH (c_3_6) properties("replication_num"="3") ;
INSERT INTO t1 VALUES (NULL, 1, "1970-01-01 08:00:00", 55932683933601.597, 11, 0.0056837147825, 20, 8424298.268498612409, true, true, 0, hll_hash("0Te"), "1970-01-01 08:00:00", 0.7309351406096245, false, 0.7309351406096245, 1), (7, 6, "1970-01-01 08:00:00", 93342467605974.954, 10, 0.0048248981631, 0, NULL, true, NULL, 0, hll_hash("CGlg"), "1970-01-01 08:00:00", 0.7309351406096245, true, 0.5550394014113348, 1), (6, 3, "1970-01-01 08:00:00", 89578924704121.617, 6, 0.0082357939878, 10, NULL, true, NULL, 3, hll_hash("b"), "1970-01-01 08:00:00", 0.8994270720003754, false, 0.5550394014113348, 6);
INSERT INTO t0 VALUES (0.00000, 4.45777, NULL, 1399899994, 0.2, 13, 5, 6999911757516781460.347, "1970-01-01 08:00:00", "xpzao8", 6, "4", NULL, false, 0.9740695738560023, hll_hash("E"), "1970-01-01", hll_hash("2"), 1.399899994E9);
INSERT INTO t2 VALUES ("1970-01-01", "4", 0.025878167339220094, true, "1970-01-01");
INSERT INTO t2 VALUES ("1970-01-01", "x", 0.060195378909065672, true, "1970-01-01"), ("1970-01-01", "j", 0.058761714889912804, false, "1970-01-01");
INSERT INTO t0 VALUES (0.00071, 4.54424, 4, -1741827985, 0.0, 3, 8, 2022619856412007139.607, "1970-01-01 08:00:00", "Si", 26, "", false, false, 0.46221160283201146, hll_hash("MT20"), "1970-01-01", hll_hash(""), 0.5550394014113348);
INSERT INTO t2 VALUES ("1970-01-01", "", 0.004181456262325298, true, "1970-01-01");
INSERT INTO t1 VALUES (2, 7, NULL, 64697770577171.054, 0, 0.0049790436620, 24, 8857975.495385287788, true, false, 1, hll_hash("Pmr8faxTM"), "1970-01-01 08:00:00", 0.32281314418335716, false, 0.51405015136195, 1);
INSERT INTO t2 VALUES ("1970-01-01", NULL, 0.040606584227191129, NULL, "1970-01-01"), ("1970-01-01", "XeJj4", 0.076005502477479462, false, "1970-01-01");
INSERT INTO t1 VALUES (7, 8, "1970-01-01 08:00:00", 24007864899805.525, 7, 0.0019379425353, 11, 82970.674190709609, true, true, 5, hll_hash("d3BqtS"), "1970-01-01 08:00:00", 0.02316683248326923, false, 0.3383206635000994, 4), (3, 6, NULL, 77750899895698.166, NULL, 0.0095618146739, 1, 3632161.131279031958, true, true, 2, hll_hash("f"), "1970-01-01 08:00:00", 0.4751502114595064, true, 1.399899994E9, 0), (4, 5, "1970-01-01 08:00:00", 69487403018150.072, 7, 0.0085325815394, 31, 4312455.763106864874, false, true, 4, hll_hash(""), "1970-01-01 08:00:00", 0.6844353663333892, true, 0.3383206635000994, 3);
-- SELECT DISTINCT t0.c_0_0 FROM  t0 FULL OUTER JOIN t1 ON t0.c_0_11 = t1.c_1_7 AND t0.c_0_11 != t1.c_1_7;
-- cardinality: 3
-- SELECT DISTINCT * FROM (SELECT DISTINCT t0.c_0_0 FROM  t0 FULL JOIN t1 ON t0.c_0_11 = t1.c_1_7 AND t0.c_0_11 != t1.c_1_7 WHERE (t0.c_0_14)  BETWEEN (t0.c_0_14) AND (POWER(CAST(t1.c_1_15 AS FLOAT ) , -6.6977298E7) )  UNION ALL SELECT DISTINCT t0.c_0_0 FROM  t0 FULL JOIN t1 ON t0.c_0_11 = t1.c_1_7 AND t0.c_0_11 != t1.c_1_7 WHERE (NOT ((t0.c_0_14)  BETWEEN (t0.c_0_14) AND (POWER(CAST(t1.c_1_15 AS FLOAT ) , -6.6977298E7) ) )) UNION ALL SELECT DISTINCT t0.c_0_0 FROM  t0 FULL OUTER JOIN t1 ON t0.c_0_11 = t1.c_1_7 AND t0.c_0_11 != t1.c_1_7 WHERE (((t0.c_0_14)  BETWEEN (t0.c_0_14) AND (POWER(CAST(t1.c_1_15 AS FLOAT ) , -6.6977298E7) ) ) IS NULL)) t;
-- cardinality: 1

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - branch-2.1 Release
 - mysql> select current_version();
+---------------------------------+
| current_version()               |
+---------------------------------+
| SQLANCER_02_RELEASE cee9fdd |
+---------------------------------+

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3252

## Problem Summary(Required) ：

NullableColumn's null flag should not be rewritten

### CODE REVIEW

File: be/src/exprs/vectorized/binary_function.h [hunk 1]
```diff
@@ -208,30 +208,27 @@ class CheckOutputBinaryFunction {
                 data = data_result;
             }
 
-            NullColumnPtr nulls;
             if (data->is_nullable()) {
-                nulls = ColumnHelper::as_raw_column<NullableColumn>(data)->null_column();
-            } else {
-                nulls = RunTimeColumnType<TYPE_NULL>::create();
-                nulls->resize(data->size());
+                return data_result;
```
Comment (stdpain) [inline be/src/exprs/vectorized/binary_function.h:212]: we need check data column

File: be/src/exprs/vectorized/binary_function.h [hunk 2]
```diff
@@ -208,30 +208,27 @@ class CheckOutputBinaryFunction {
...
+            auto& row_real_data = ColumnHelper::cast_to_raw<ResultType>(real_data)->get_data();
 
             for (size_t i = 0; i < data->size(); ++i) {
-                ns[i] = NULL_OP::template apply<RunTimeCppType<ResultType>, RunTimeCppType<ResultType>>(r1[i]);
+                null_flag->get_data()[i] =
+                        NULL_OP::template apply<RunTimeCppType<ResultType>, RunTimeCppType<ResultType>>(
+                                row_real_data[i]);
             }
 
-            if (!data->is_nullable()) {
```
Comment (stdpain) [inline be/src/exprs/vectorized/binary_function.h:227]: we should call `down_cast<NullableColumn*>(data.get())->update_has_null();`

File: be/src/exprs/vectorized/binary_function.h [hunk 3]
```diff
@@ -208,30 +208,27 @@ class CheckOutputBinaryFunction {
...
             }
-            auto* ns = nulls->get_data().data();
+            NullColumnPtr null_flag = RunTimeColumnType<TYPE_NULL>::create();
+            null_flag->resize(data->size());
 
             const ColumnPtr& real_data = FunctionHelper::get_real_data_column(data);
-            auto& r1 = ColumnHelper::cast_to_raw<ResultType>(real_data)->get_data();
+            auto& row_real_data = ColumnHelper::cast_to_raw<ResultType>(real_data)->get_data();
 
             for (size_t i = 0; i < data->size(); ++i) {
```
[the remark this answers is not in the document]
Comment (liuyehcf) [reply be/src/exprs/vectorized/binary_function.h:224]: ok

File: be/src/exprs/vectorized/binary_function.h [hunk 4]
```diff
@@ -208,25 +208,31 @@ class CheckOutputBinaryFunction {
...
-            auto& r1 = ColumnHelper::cast_to_raw<ResultType>(real_data)->get_data();
+
+            // Avoid calling virtual fuction `get_data` in for loop
+            auto& row_null_flags = null_flags->get_data().data();
+            auto& row_real_data = ColumnHelper::cast_to_raw<ResultType>(real_data)->get_data();
 
             for (size_t i = 0; i < data->size(); ++i) {
-                ns[i] = NULL_OP::template apply<RunTimeCppType<ResultType>, RunTimeCppType<ResultType>>(r1[i]);
+                // DO NOT overwrite null flag if it is already set
+                if (row_null_flags[i]) {
```
Comment (stdpain) [inline be/src/exprs/vectorized/binary_function.h:226]: we could remove condition in loop
```
row_null_flags[i] &= NULL_OP::apply(...
```

**Review by stdpain — DISMISSED** [review]

**Review by satanson — DISMISSED** [review]

**Comment by kangkaisen** [comment]
please add UT

**Comment by liuyehcf** [comment]
run starrocks_clang-format

**Comment by liuyehcf** [comment]
run starrocks_be_unittest

**Pull request merged**